File: 56-merge-intervals/merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> ans ;
        int b = -1,a = -1; // before, after.

        sort(intervals.begin(), intervals.end());

        for(int i=0; i<intervals.size() ; i++) {
            if (b == -1) b = i;
            if (a == -1) a = i;
            
            b = (intervals[b][0] < intervals[i][0])? b : i;
            a = (intervals[a][1] > intervals[i][1])? a : i;

            if (i+1 == intervals.size()) {
                ans.push_back({intervals[b][0], intervals[a][1]});
                break;
            }

            // split and push
            if (intervals[a][1] < intervals[i+1][0]) {
                ans.push_back({intervals[b][0], intervals[a][1]});
                b = -1;
                a = -1;
            }
        }

        return ans;
    }
};

File: 56-merge-intervals/merge-intervals.java
class Solution {

    // [2,3], [3,4], []
    // [2,4], [1,3], []
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0], b[0]));
        List<int[]> ans = new ArrayList<>();

        int l = -1;
        int r = -1;
        for(int i = 0 ; i < intervals.length; i++) {
            if (l == -1) {
                l = intervals[i][0];
                r = intervals[i][1];
            } else {
                if (intervals[i][0] <= r) {
                    r = Math.max(intervals[i][1], r);
                } else {
                    ans.add(new int[]{l,r});
                    l = intervals[i][0];
                    r = intervals[i][1];
                }
            }

            if (i == intervals.length-1) {
                ans.add(new int[]{l,r});
            }
        }

        return ans.toArray(new int[ans.size()][2]);
    }
}